model.py

```diff
@@ -217,6 +217,15 @@ def forward(self, trg_batch: Tensor, h_encoder: Tensor) -> Tensor:
                 outputs.append(output)
         return torch.cat(outputs, 0)
 
+    def _batched_nearest_neighbor(self, x: Tensor, size: int) -> Tensor:
```

Comment:
Couldn't this be vectorized? It would  give us quite a speed up and avoid eval being the bottleneck.

I suggest looking at: https://discuss.pytorch.org/t/k-nearest-neighbor-in-pytorch/59695

For batches you might be able to use BMM. We can look at that together

Reply:
Moreover, is there a reason this is in the model rather than stand alone function? We may want to use this on other models or even in other contexts. I suggest moving this (unless there is a good reason) to utils.get_nearest_neighbor and calling it with the model members if necessary

Reply:
If I vectorize this for vMF decoding, it will OOM because vMF training can run with high batch size but if I try to do a nearest neighbor search, it takes up orders of magnitude more memory so I have to mini mini batch it.

Reply:
Sachin's paper uses a batch size of 64. Yesterday I think you mentioned a batch size of 8192, maybe we should stick to his hyperparams for now.

If you do decide to use that higher batch size, why not use 2 separate batch sizes if that's the issue instead of this? This seems like a clumsy solution to a non-essential issue